src/classes/frDonation.cls

```diff
@@ -144,16 +150,75 @@ public class frDonation extends frModel {
+            Boolean.valueOf(request.get('opportunityContactMappingDisabled')): true;
+        if (!mappingDisabled) {
+            String funraiseId = String.valueOf(request.get('id'));
+            String supporterId = String.valueOf(request.get('donorId'));
+            
+            Set<String> relatedSupporterFunraiseIds = new Set<String>{supporterId};
+            String fundraiserId = request.containsKey('fundraiserId') ? String.valueOf(request.get('fundraiserId')) : null;
+            if(String.isNotBlank(fundraiserId)) {
+                relatedSupporterFunraiseIds.add(fundraiserId);
+            }
+            String teamCaptainId = request.containsKey('teamCaptainId') ? String.valueOf(request.get('teamCaptainId')) : null;
+            if(String.isNotBlank(teamCaptainId)) {
+                relatedSupporterFunraiseIds.add(teamCaptainId);
+            }
+            Map<String, Contact> relatedContacts = new Map<String, Contact>();
+            for(Contact relatedContact : [SELECT Id, fr_Id__c, 
+                                          (SELECT Id, Role, ContactId FROM OpportunityContactRoles WHERE OpportunityId = :getOpportunityId()) 
+                                          from Contact WHERE fr_Id__c IN :relatedSupporterFunraiseIds]) {
+                relatedContacts.put(relatedContact.fr_Id__c, relatedContact);
+            }
+            
+            List<OpportunityContactRole> newContactRoles = new List<OpportunityContactRole>();
+            OpportunityContactRole donorRole = createRole(relatedContacts.get(supporterId), OPP_ROLE_DONOR);
+            if(donorRole != null) newContactRoles.add(donorRole);
+            
+            if(String.isNotBlank(fundraiserId)) {
+                if(!relatedContacts.containsKey(fundraiserId)) {
+                    insert new Error__c(Error__c =
+                                        'Failed to find related record: Contact with funraise id ' 
+                                        + fundraiserId+ 
+                                        ' for opportunity contact role '+ OPP_ROLE_FUNDRAISER +' for opportunity with funraise id ' 
+                                        + funraiseId);
+                } else {
+                    OpportunityContactRole fundraiserRole = createRole(relatedContacts.get(fundraiserId), OPP_ROLE_FUNDRAISER);
+                    if(fundraiserRole != null) newContactRoles.add(fundraiserRole);
+                }
+            }
+            if(String.isNotBlank(teamCaptainId)) {
+                if(!relatedContacts.containsKey(teamCaptainId)) {
+                    insert new Error__c(Error__c =
+                                        'Failed to find related record: Contact with funraise id ' 
+                                        + teamCaptainId+ 
+                                        ' for opportunity contact role '+ OPP_ROLE_TEAM_CAPTAIN +' for opportunity with funraise id ' 
+                                        + funraiseId);
```

Comment:
Possibly methodize this error message if you need to add a similar one for `supporterId` also